### PR TITLE
Make all backends optional

### PR DESCRIPTION
**Pull request by sethfowler**
People who are developing extensions won't necessarily want to build all of the backends - for example, they may want to avoid building anything unnecessary on their continuous integration infrastructure. (That's my motivation in writing this PR.)

What this PR does is:
- Add flags to enable or disable the backends. They all default to ON.
- Use the existing infrastructure to avoid including BMV2-specific GTests if the BMV2 backend is not enabled.

With those changes in place, things "just work". Disabling all the backends will give you a perfectly working test suite consisting of two tests, cpplint and gtest. =)

As a bonus, I've also including a commit in this PR to make `libboost-graph` available on Travis so we will build the `p4c-graphs` backend there.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -30,6 +30,10 @@ set (CMAKE_USE_RELATIVE_PATHS 1)
 
 OPTION (ENABLE_DOCS "Build the documentation" OFF)
 OPTION (ENABLE_P4RUNTIME_TO_PD "Build the P4Runtime PD Generator" ON)
+OPTION (ENABLE_BMV2 "Build the BMV2 backend (required for the full test suite)" ON)
+OPTION (ENABLE_EBPF "Build the EBPF backend (required for the full test suite)" ON)
+OPTION (ENABLE_EBPF "Build the P4Test backend (required for the full test suite)" ON)
```
Comment (ChrisDodd) [inline CMakeLists.txt]: typo here? EBPF vs P4Test

**Comment by ChrisDodd** [comment]
Should there also be something in `foreach (ext` loop around line 164 of CMakeLists.txt to create an `ENABLE_` flag for each subdirectory in extensions?  I guess it requires string manipulation to turn the directory name into a valid identifier.

File: CMakeLists.txt (hunk 1 above)
Comment (sethfowler) [reply CMakeLists.txt]: Doh! You're right. I'll fix that.

**Comment by sethfowler** [comment]
That does sound like a good improvement. Let me see if it's a simple thing to do.

**Comment by sethfowler** [comment]
I honestly expected that to be harder than it was. OK, we now automatically generate an option to disable each extension we find.

File: CMakeLists.txt [hunk 2]
```diff
@@ -193,6 +197,13 @@ file (GLOB p4c_extensions RELATIVE ${P4C_SOURCE_DIR}/extensions ${P4C_SOURCE_DIR
...
 foreach (ext ${p4c_extensions})
   if (EXISTS ${P4C_SOURCE_DIR}/extensions/${ext}/CMakeLists.txt)
+    # Generate an option that makes it possible to disable this extension.
+    string(MAKE_C_IDENTIFIER ${ext} EXT_AS_IDENTIFIER)
+    string(TOUPPER ${EXT_AS_IDENTIFIER} EXT_AS_OPTION_NAME)
+    string(CONCAT ENABLE_EXT_OPTION "ENABLE_" ${EXT_AS_OPTION_NAME})
+    string(CONCAT EXT_HELP_TEXT "Build the " ${ext} " backend")
+    OPTION (${ENABLE_EXT_OPTION} ${EXT_HELP_TEXT} ON)
+
     add_subdirectory (extensions/${ext})
```
Comment (ChrisDodd) [inline CMakeLists.txt]: should this be wrapped in `if (${ENABLE_EXT_OPTION})`?

File: CMakeLists.txt [hunk 3]
```diff
@@ -30,6 +30,10 @@ set (CMAKE_USE_RELATIVE_PATHS 1)
 
 OPTION (ENABLE_DOCS "Build the documentation" OFF)
 OPTION (ENABLE_P4RUNTIME_TO_PD "Build the P4Runtime PD Generator" ON)
+OPTION (ENABLE_BMV2 "Build the BMV2 backend (required for the full test suite)" ON)
+OPTION (ENABLE_EBPF "Build the EBPF backend (required for the full test suite)" ON)
+OPTION (ENABLE_P4TEST "Build the P4Test backend (required for the full test suite)" ON)
+OPTION (ENABLE_P4C_GRAPHS "Build the p4c-graphs backend" ON)
```
Comment (cc10512) [inline CMakeLists.txt]: shouldn't you remove these if you've enabled them below now?

File: CMakeLists.txt [hunk 4]
```diff
@@ -202,19 +213,25 @@ add_subdirectory (midend)
 add_subdirectory (control-plane)
 
 
+add_subdirectory (test)
+
+if (ENABLE_BMV2)
```
Comment (cc10512) [inline CMakeLists.txt]: and these are now also redundant.

File: CMakeLists.txt (hunk 2 above)
Comment (sethfowler) [reply CMakeLists.txt]: Argh, yes, it should.

File: CMakeLists.txt (hunk 3 above)
Comment (sethfowler) [reply CMakeLists.txt]: I don't understand... the foreach loop is for extensions. These are built-in backends. Am I missing something?

**Comment by sethfowler** [comment]
OK, the version I just pushed adds the missing `if()` Chris pointed out, so disabling extensions actually works. I checked that this actually works locally.

I've also improved the formatting a little bit in `CMakeLists.txt`.

File: CMakeLists.txt (hunk 3 above)
Comment (cc10512) [reply CMakeLists.txt]: You're right. I read is as all backends being captured in that loop. Sorry about that.

Comment (sethfowler) [reply CMakeLists.txt]: No worries. Thanks for reviewing so closely. =)

Comment (ChrisDodd) [reply CMakeLists.txt]: Though I guess you /could/ have a loop over the backends directory to create all of these.  I guess the help messages would then be less informational (unless there was a way of pulling the message from a file in the directory in question...)

**Pull request merged**